### ISSUE TITLE
#send_msg is calling #notify_readable causing messages to stack up

### ISSUE THREAD

**Issue by jgyllen** [andrewvc/em-zeromq#20]
https://github.com/andrewvc/em-zeromq/blob/master/lib/em-zeromq/socket.rb#L81

I have a DEALER-ROUTER setup. Inbound messages (from the dealer) being processed by the router usually involve sending messages to one or more dealer peers. When the router calls #send_msg, that method in turn calls #notify_readable which checks if there are new messages on the socket and if so fires #on_readable. This causes the #on_readable calls to stack up (in a LIFO fashion) and the outer ones won't be completed until the inbound message flow ceases and it can wind back down the stack.

Basically, if I have a fairly high rate of incoming messages on the router and those trigger multiple outbound messages, only the first outbound message gets sent until the rate decreases and it has time to complete processing of any pending messages.

Is this a bug? Or is it by design? If so, what's the reasoning behind it and how can the driver be configured to avoid this issue when using the above pattern?

**Comment by andrewvc** [comment]
@schmurfy looks like this was introduced here: https://github.com/andrewvc/em-zeromq/commit/9e4b07148e51d7208c3ba1233ef5b1712313edcf

Seems like it's here for a reason, though we'll clearly have to fix this bug. Any ideas?

**Comment by jgyllen** [comment]
We tried commenting the line in a fork which prevents the stacking of messages. But using the same fork in another app in the same system causes a REQ-REP pair to fail.

**Comment by andrewvc** [comment]
I'm thinking the right thing to do is to leave the notify_readable() in, but instead of immediately invoking on_readable directly in notify_readable use EM::next_tick to push it onto the reactor queue.

That should be prevent the stack from deepening and shouldn't affect performance noticeably.

**Comment by jgyllen** [comment]
Sure, but wouldn't #on_readable be fired by EventMachine in the next tick anyway if there is new data on the socket?

**Comment by jgyllen** [comment]
I mean #notify_readable

**Comment by andrewvc** [comment]
@jgyllen That's a good point, I'm guessing that there's a reason that might not be the case given that @schmurfy added that call intentionally. Hopefully he'll be able to shed some light on the issue.

**Comment by jgyllen** [comment]
Yeah, that's what I was thinking too. Thanks for pitching in.

**Comment by schmurfy** [comment]
From what I remember I found a use case where a message was in fact waiting to be read but notify_readable was never called and since the socket is edge triggered the process was stuck in this state, I did a lot of testing trying to reproduce the issue at the C level but never got to reproduce it so I suppose this is related to the ruby part but I do not know where :/

I think the issue happens only when the REQ/REP or XREQ/XREP sockets are inside the same process though so if it cause problems the call to notify_readable could be removed, maybe some timing issue.

**Comment by andrewvc** [comment]
Thanks for the confirmation schmurfy. I'll move it to next_tick, that should clear up the issue (it'll still get triggered, but the stack will have a chance to unwind)

**Comment by jgyllen** [comment]
@andrewvc Good idea. Will you be pushing out a new gem?

**Comment by schmurfy** [comment]
I pushed the change, can you git it a try ?
If it works for you I will push a new release out.

**Comment by jgyllen** [comment]
Thanks @schmurfy. Will give it a try and let you know.

**Comment by jgyllen** [comment]
@schmurfy We tried the fix and it works. Please let me know when you have made a new release. Thanks!

**Comment by schmurfy** [comment]
0.3.1 just released !
